feat: implement `glimpse()` for `zoomed_df`

IndrajeetPatil opened the pull request:
Closes #1003

TSchiefer commented:
I like this implementation, thank you for taking this on.
If I may just give one comment: in case key columns get lost in transformations, they will currently still appear in the glimpse-output. Maybe you could check them against `col_tracker_zoomed(x)`?

IndrajeetPatil commented:
@TSchiefer Great catch! 

Fixed it (and added a test):

``` r
library(dm)

dm_nycflights13() %>% dm_zoom_to(weather) %>% select(-origin) %>% glimpse()
#> dm of 5 tables: `airlines`, `airports`, `flights`, `planes`, `weather`
#> 
#> Table: `weather`
#> Primary key: (`time_hour`)
#> Rows: 144
#> Columns: 14
#> $ year       <int> 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013,…
#> $ month      <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,…
#> $ day        <int> 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,…
#> $ hour       <int> 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 1…
#> $ temp       <dbl> 41.00, 39.02, 39.02, 39.92, 41.00, 41.00, 39.92, 41.00, 42.…
#> $ dewp       <dbl> 32.00, 30.02, 28.94, 26.96, 26.06, 26.06, 24.98, 24.98, 24.…
#> $ humid      <dbl> 70.08, 69.86, 66.85, 59.50, 54.97, 54.97, 54.81, 52.56, 48.…
#> $ wind_dir   <dbl> 230, 210, 230, 270, 320, 300, 280, 330, 330, 320, 320, 330,…
#> $ wind_speed <dbl> 8.05546, 9.20624, 6.90468, 5.75390, 6.90468, 12.65858, 6.90…
#> $ wind_gust  <dbl> NA, NA, NA, NA, NA, 20.71404, 17.26170, NA, NA, 26.46794, N…
#> $ precip     <dbl> 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,…
#> $ pressure   <dbl> 1024.6, 1025.9, 1026.9, 1027.5, 1028.2, 1029.0, 1030.0, 103…
#> $ visib      <dbl> 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,…
#> $ time_hour  <dttm> 2013-01-10 00:00:00, 2013-01-10 01:00:00, 2013-01-10 02:00…
```

<sup>Created on 2022-06-23 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

krlmlr commented on tests/testthat/_snaps/dm.md:
```diff
@@ -272,3 +272,78 @@
+      dm of 5 tables: `airlines`, `airports`, `flights`, `planes`, `weather`
+      
+      Table: `weather`
+      Primary key: (`time_hour`)
```
The pk should be removed here:

``` r
library(conflicted)
library(dm)

dm_nycflights13() %>%
  dm_select_tbl(weather) %>%
  dm_select(weather, -origin) %>%
  dm_get_all_pks()
#> # A tibble: 0 × 2
#> # … with 2 variables: table <chr>, pk_col <keys>
```

<sup>Created on 2022-06-24 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

IndrajeetPatil replied:
Thanks. Definitely a major issue to attend to.

Better?

``` r
library(dm)

dm_nycflights13() %>%
  dm_select_tbl(weather) %>%
  dm_select(weather, -origin) %>%
  glimpse()
#> dm of 1 tables: `weather`
#> 
#> Table: `weather`
#> Rows: 144
#> Columns: 14
#> $ year       <int> 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013,…
#> $ month      <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,…
#> $ day        <int> 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,…
#> $ hour       <int> 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 1…
#> $ temp       <dbl> 41.00, 39.02, 39.02, 39.92, 41.00, 41.00, 39.92, 41.00, 42.…
#> $ dewp       <dbl> 32.00, 30.02, 28.94, 26.96, 26.06, 26.06, 24.98, 24.98, 24.…
#> $ humid      <dbl> 70.08, 69.86, 66.85, 59.50, 54.97, 54.97, 54.81, 52.56, 48.…
#> $ wind_dir   <dbl> 230, 210, 230, 270, 320, 300, 280, 330, 330, 320, 320, 330,…
#> $ wind_speed <dbl> 8.05546, 9.20624, 6.90468, 5.75390, 6.90468, 12.65858, 6.90…
#> $ wind_gust  <dbl> NA, NA, NA, NA, NA, 20.71404, 17.26170, NA, NA, 26.46794, N…
#> $ precip     <dbl> 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,…
#> $ pressure   <dbl> 1024.6, 1025.9, 1026.9, 1027.5, 1028.2, 1029.0, 1030.0, 103…
#> $ visib      <dbl> 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,…
#> $ time_hour  <dttm> 2013-01-10 00:00:00, 2013-01-10 01:00:00, 2013-01-10 02:00…
```

<sup>Created on 2022-06-24 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

TSchiefer replied:
Sorry to get involved again, but I think this was a misunderstanding: https://github.com/cynkra/dm/pull/1161/commits/77fba8ea93c43365950d03130f4b6ade955aa8de does not really change the behavior of the function, since for unzoomed dm's the result for `dm_get_pk_impl(x, table_name)` would be an empty list anyway.

I believe what Kirill meant was, that if one column that is part of a compound key -- be it PK or FK -- gets lost for a zoomed dm, we want to drop that key altogether, as it would be done in case the zoomed table would be updated or inserted.

``` r
library(dm)

# should return this:
dm_nycflights13() %>% dm_zoom_to(weather) %>% select(-origin) %>% glimpse()
#> dm of 5 tables: `airlines`, `airports`, `flights`, `planes`, `weather`
#> 
#> Table: `weather`
#> Rows: 144
#> Columns: 14
#> $ year       <int> 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013,…
#> $ month      <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,…
#> $ day        <int> 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,…
#> $ hour       <int> 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 1…
#> $ temp       <dbl> 41.00, 39.02, 39.02, 39.92, 41.00, 41.00, 39.92, 41.00, 42.…
#> $ dewp       <dbl> 32.00, 30.02, 28.94, 26.96, 26.06, 26.06, 24.98, 24.98, 24.…
#> $ humid      <dbl> 70.08, 69.86, 66.85, 59.50, 54.97, 54.97, 54.81, 52.56, 48.…
#> $ wind_dir   <dbl> 230, 210, 230, 270, 320, 300, 280, 330, 330, 320, 320, 330,…
#> $ wind_speed <dbl> 8.05546, 9.20624, 6.90468, 5.75390, 6.90468, 12.65858, 6.90…
#> $ wind_gust  <dbl> NA, NA, NA, NA, NA, 20.71404, 17.26170, NA, NA, 26.46794, N…
#> $ precip     <dbl> 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,…
#> $ pressure   <dbl> 1024.6, 1025.9, 1026.9, 1027.5, 1028.2, 1029.0, 1030.0, 103…
#> $ visib      <dbl> 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,…
#> $ time_hour  <dttm> 2013-01-10 00:00:00, 2013-01-10 01:00:00, 2013-01-10 02:00…

# and in case of renaming:
dm_nycflights13() %>% dm_zoom_to(weather) %>% rename(origin_location = origin) %>% glimpse()
#> dm of 5 tables: `airlines`, `airports`, `flights`, `planes`, `weather`
#> 
#> Table: `weather`
#> Primary key: (`origin_location`, `time_hour`)
#> Rows: 144
#> Columns: 15
#> $ origin_location <chr> "EWR", "EWR", "EWR", "EWR", "EWR", "EWR", "EWR", "EWR"…
#> $ year            <int> 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, …
#> $ month           <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, …
#> $ day             <int> 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10…
#> $ hour            <int> 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, …
#> $ temp            <dbl> 41.00, 39.02, 39.02, 39.92, 41.00, 41.00, 39.92, 41.00…
#> $ dewp            <dbl> 32.00, 30.02, 28.94, 26.96, 26.06, 26.06, 24.98, 24.98…
#> $ humid           <dbl> 70.08, 69.86, 66.85, 59.50, 54.97, 54.97, 54.81, 52.56…
#> $ wind_dir        <dbl> 230, 210, 230, 270, 320, 300, 280, 330, 330, 320, 320,…
#> $ wind_speed      <dbl> 8.05546, 9.20624, 6.90468, 5.75390, 6.90468, 12.65858,…
#> $ wind_gust       <dbl> NA, NA, NA, NA, NA, 20.71404, 17.26170, NA, NA, 26.467…
#> $ precip          <dbl> 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, …
#> $ pressure        <dbl> 1024.6, 1025.9, 1026.9, 1027.5, 1028.2, 1029.0, 1030.0…
#> $ visib           <dbl> 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10…
#> $ time_hour       <dttm> 2013-01-10 00:00:00, 2013-01-10 01:00:00, 2013-01-10 …
```

<sup>Created on 2022-06-27 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

And the same behavior for FKs.

For PKs you could achieve that for example with:
```
  # anticipate that some key columns could have been removed by the user
  if (is_zoomed(x)) {
    tracked_cols <- col_tracker_zoomed(x)
    pk <- if (all(pk[[1]] %in% tracked_cols)) {
      list(unname(prep_recode(tracked_cols)[pk[[1]]]))
    } else {
      NULL
    }
  }
```
but please feel free to change the implementation.

IndrajeetPatil replied:
Thanks for the clarification, Tobias! I was definitely confused about the expected output, so the example code really helped.

I've implemented this change for primary keys, but still need to handle the foreign keys.

IndrajeetPatil replied:
Is there an equivalent of `update_zoomed_pk()` for foreign keys?

TSchiefer replied:
Good that you ask: at first I thought it is `update_zoomed_outgoing()` in R/zoom.R, but this function is called nowhere (and is also not giving the correct result).
The correct operation is done in `dm_insert_zoomed_outgoing_fks()`, in the calculation of the variable `new_out_keys`.
I would suggest to extract the code in this assignment and replace the unused function `update_zoomed_outgoing()` with it, if this makes sense.

TSchiefer replied:
My apologies, actually `update_zoomed_outgoing()` is being called in `dm_update_zoomed()`. Maybe you can make use of it; but actually using the code in `dm_insert_zoomed_outgoing_fks()` still seems easiest to me. 
In any case I am not anymore of the opinion that `update_zoomed_outgoing()` should be replaced.

krlmlr reviewed (commented):
Thanks, one minor (major?) nit.

krlmlr commented on R/dm.R:
```diff
@@ -732,56 +732,144 @@ as.list.zoomed_dm <- function(x, ...) {
+  all_fks <- dm_get_all_fks_impl(x)
+
+  fk <- all_fks %>%
+    filter(child_table == table_name) %>%
```
I try to be explicit, even if we control this data structure. No need to act now.

```suggestion
    filter(child_table == !!table_name) %>%
```

krlmlr commented on R/dm.R:
```diff
@@ -732,56 +732,144 @@ as.list.zoomed_dm <- function(x, ...) {
+  # return early if there are no foreign keys, since there is nothing to print here
+  if (!is_zoomed(x)) {
+    fk_df <- dm_get_all_fks(x)
+
+    if (nrow(fk_df) == 0L) {
+      return(invisible(x))
     }
-    glimpse(table, width = width, ...)
-  })
-  invisible(x)
+  }
```
What happens if we omit this shortcut? If we really need it, should we move the call to `dm_get_all_fks_impl(x)` up?

krlmlr commented on tests/testthat/_snaps/dm.md:
```diff
@@ -271,4 +271,102 @@
+      dm of 5 tables: `airlines`, `airports`, `flights`, `planes`, `weather`
+      
+      Table: `airports`
+      Primary key: (`faa`)
```
Should we use parens only for compound keys (which are pretty rare)?

```suggestion
      Primary key: `faa`
```

IndrajeetPatil replied:
Good idea!

I have implemented this, both for primary and foreign keys (LHS and RHS entries).

krlmlr commented on tests/testthat/_snaps/dm.md:
```diff
@@ -271,4 +271,102 @@
+      
+      Table: `flights`
+      4 outgoing foreign key(s):
+        (`carrier`) -> (`airlines$carrier`) no_action
```
Same here.

```suggestion
        `carrier` -> `airlines$carrier` no_action
```

krlmlr reviewed (commented):
Thanks, let's do another round.

IndrajeetPatil commented:
Thank you, both, for helpful pointers! 

I think most of the suggestions made are now implemented in this PR:

## Primary keys

``` r
library(dm)

# if any primary key has been removed, no primary key is displayed
dm_nycflights13() %>% dm_zoom_to(weather) %>% select(-origin) %>% glimpse()
#> dm of 5 tables: `airlines`, `airports`, `flights`, `planes`, `weather`
#> 
#> Table: `weather`
#> Rows: 144
#> Columns: 14
#> $ year       <int> 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013,…
#> $ month      <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,…
#> $ day        <int> 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,…
#> $ hour       <int> 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 1…
#> $ temp       <dbl> 41.00, 39.02, 39.02, 39.92, 41.00, 41.00, 39.92, 41.00, 42.…
#> $ dewp       <dbl> 32.00, 30.02, 28.94, 26.96, 26.06, 26.06, 24.98, 24.98, 24.…
#> $ humid      <dbl> 70.08, 69.86, 66.85, 59.50, 54.97, 54.97, 54.81, 52.56, 48.…
#> $ wind_dir   <dbl> 230, 210, 230, 270, 320, 300, 280, 330, 330, 320, 320, 330,…
#> $ wind_speed <dbl> 8.05546, 9.20624, 6.90468, 5.75390, 6.90468, 12.65858, 6.90…
#> $ wind_gust  <dbl> NA, NA, NA, NA, NA, 20.71404, 17.26170, NA, NA, 26.46794, N…
#> $ precip     <dbl> 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,…
#> $ pressure   <dbl> 1024.6, 1025.9, 1026.9, 1027.5, 1028.2, 1029.0, 1030.0, 103…
#> $ visib      <dbl> 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,…
#> $ time_hour  <dttm> 2013-01-10 00:00:00, 2013-01-10 01:00:00, 2013-01-10 02:00…

# anticipate primary keys being renamed by users
dm_nycflights13() %>% dm_zoom_to(weather) %>% rename(origin_location = origin) %>% glimpse()
#> dm of 5 tables: `airlines`, `airports`, `flights`, `planes`, `weather`
#> 
#> Table: `weather`
#> Primary key: (`origin_location`, `time_hour`)
#> Rows: 144
#> Columns: 15
#> $ origin_location <chr> "EWR", "EWR", "EWR", "EWR", "EWR", "EWR", "EWR", "EWR"…
#> $ year            <int> 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, …
#> $ month           <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, …
#> $ day             <int> 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10…
#> $ hour            <int> 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, …
#> $ temp            <dbl> 41.00, 39.02, 39.02, 39.92, 41.00, 41.00, 39.92, 41.00…
#> $ dewp            <dbl> 32.00, 30.02, 28.94, 26.96, 26.06, 26.06, 24.98, 24.98…
#> $ humid           <dbl> 70.08, 69.86, 66.85, 59.50, 54.97, 54.97, 54.81, 52.56…
#> $ wind_dir        <dbl> 230, 210, 230, 270, 320, 300, 280, 330, 330, 320, 320,…
#> $ wind_speed      <dbl> 8.05546, 9.20624, 6.90468, 5.75390, 6.90468, 12.65858,…
#> $ wind_gust       <dbl> NA, NA, NA, NA, NA, 20.71404, 17.26170, NA, NA, 26.467…
#> $ precip          <dbl> 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, …
#> $ pressure        <dbl> 1024.6, 1025.9, 1026.9, 1027.5, 1028.2, 1029.0, 1030.0…
#> $ visib           <dbl> 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10…
#> $ time_hour       <dttm> 2013-01-10 00:00:00, 2013-01-10 01:00:00, 2013-01-10 …
```

## Foreign keys

```r
# if any foreign key has been removed, no foreign key is displayed
dm_nycflights13() %>% dm_zoom_to(flights) %>% select(-origin) %>% glimpse()
#> dm of 5 tables: `airlines`, `airports`, `flights`, `planes`, `weather`
#> 
#> Table: `flights`
#> Rows: 1,761
#> Columns: 18
#> $ year           <int> 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2…
#> $ month          <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1…
#> $ day            <int> 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,…
#> $ dep_time       <int> 3, 16, 450, 520, 530, 531, 535, 546, 549, 550, 553, 553…
#> $ sched_dep_time <int> 2359, 2359, 500, 525, 530, 540, 540, 600, 600, 600, 600…
#> $ dep_delay      <dbl> 4, 17, -10, -5, 0, -9, -5, -14, -11, -10, -7, -7, -7, -…
#> $ arr_time       <int> 426, 447, 634, 813, 824, 832, 1015, 645, 652, 649, 711,…
#> $ sched_arr_time <int> 437, 444, 648, 820, 829, 850, 1017, 709, 724, 703, 715,…
#> $ arr_delay      <dbl> -11, 3, -14, -7, -5, -18, -2, -24, -32, -14, -4, -33, -…
#> $ carrier        <chr> "B6", "B6", "US", "UA", "UA", "AA", "B6", "B6", "EV", "…
#> $ flight         <int> 727, 739, 1117, 1018, 404, 1141, 725, 380, 6055, 2114, …
#> $ tailnum        <chr> "N571JB", "N564JB", "N171US", "N35204", "N815UA", "N5EA…
#> $ dest           <chr> "BQN", "PSE", "CLT", "IAH", "IAH", "MIA", "BQN", "BOS",…
#> $ air_time       <dbl> 183, 191, 78, 215, 210, 149, 191, 39, 48, 36, 51, 201, …
#> $ distance       <dbl> 1576, 1617, 529, 1400, 1416, 1089, 1576, 200, 229, 184,…
#> $ hour           <dbl> 23, 23, 5, 5, 5, 5, 5, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6,…
#> $ minute         <dbl> 59, 59, 0, 25, 30, 40, 40, 0, 0, 0, 0, 0, 0, 0, 0, 5, 0…
#> $ time_hour      <dttm> 2013-01-10 23:00:00, 2013-01-10 23:00:00, 2013-01-10 0…

# anticipate foreign keys being renamed by users
dm_nycflights13() %>% dm_zoom_to(flights) %>% rename(origin_location = origin) %>% glimpse()
#> dm of 5 tables: `airlines`, `airports`, `flights`, `planes`, `weather`
#> 
#> Table: `flights`
#> 4 outgoing foreign key(s):
#>   `carrier` -> `airlines$carrier` no_action
#>   `origin_location` -> `airports$faa` no_action
#>   `tailnum` -> `planes$tailnum` no_action
#>   (`origin_location`, `time_hour`) -> (`weather$origin`, `weather$time_hour`) no…
#> Rows: 1,761
#> Columns: 19
#> $ year            <int> 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, …
#> $ month           <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, …
#> $ day             <int> 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10…
#> $ dep_time        <int> 3, 16, 450, 520, 530, 531, 535, 546, 549, 550, 553, 55…
#> $ sched_dep_time  <int> 2359, 2359, 500, 525, 530, 540, 540, 600, 600, 600, 60…
#> $ dep_delay       <dbl> 4, 17, -10, -5, 0, -9, -5, -14, -11, -10, -7, -7, -7, …
#> $ arr_time        <int> 426, 447, 634, 813, 824, 832, 1015, 645, 652, 649, 711…
#> $ sched_arr_time  <int> 437, 444, 648, 820, 829, 850, 1017, 709, 724, 703, 715…
#> $ arr_delay       <dbl> -11, 3, -14, -7, -5, -18, -2, -24, -32, -14, -4, -33, …
#> $ carrier         <chr> "B6", "B6", "US", "UA", "UA", "AA", "B6", "B6", "EV", …
#> $ flight          <int> 727, 739, 1117, 1018, 404, 1141, 725, 380, 6055, 2114,…
#> $ tailnum         <chr> "N571JB", "N564JB", "N171US", "N35204", "N815UA", "N5E…
#> $ origin_location <chr> "JFK", "JFK", "EWR", "EWR", "LGA", "JFK", "JFK", "EWR"…
#> $ dest            <chr> "BQN", "PSE", "CLT", "IAH", "IAH", "MIA", "BQN", "BOS"…
#> $ air_time        <dbl> 183, 191, 78, 215, 210, 149, 191, 39, 48, 36, 51, 201,…
#> $ distance        <dbl> 1576, 1617, 529, 1400, 1416, 1089, 1576, 200, 229, 184…
#> $ hour            <dbl> 23, 23, 5, 5, 5, 5, 5, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6…
#> $ minute          <dbl> 59, 59, 0, 25, 30, 40, 40, 0, 0, 0, 0, 0, 0, 0, 0, 5, …
#> $ time_hour       <dttm> 2013-01-10 23:00:00, 2013-01-10 23:00:00, 2013-01-10 …
```

<sup>Created on 2022-06-29 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

TSchiefer commented on R/dm.R:
```diff
@@ -732,56 +732,149 @@ as.list.zoomed_dm <- function(x, ...) {
-      cat_line(fk)
+
+  if (is_zoomed(x)) {
+    # anticipate that some key columns could have been removed by the user
+    new_fks <- update_zoomed_fks(x, table_name, col_tracker_zoomed(x))
+
+    # if any of the key columns are missing, don't display any of the keys
+    if (nrow(new_fks) < nrow(all_fks)) {
+      return(invisible(x))
     }
-    glimpse(table, width = width, ...)
-  })
-  invisible(x)
+
+    # otherwise update the foreign keys data frame
+    all_fks <- new_fks
+  }
```
Now we're really almost there, another minor misunderstanding: we just want to remove a composite key altogether, in case on of its columns is missing. But the FKs that will still be used once the user calls `dm_insert_zoomed()` or `dm_update_zoomed()` should actually be displayed.
So the if-condition regarding `nrow(new_fks) < nrow(all_fks)` would need to be removed.

Suggestion:
```
all_fks <- if (is_zoomed(x)) {
    # anticipate that some key columns could have been removed/renamed by the user
    update_zoomed_fks(x, table_name, col_tracker_zoomed(x))
  } else {
    dm_get_all_fks_impl(x)
  }
  ```

IndrajeetPatil replied:
Makes sense.

I think it should be in line with what you expect now:

``` r
library(dm)

dm_nycflights13() %>% dm_zoom_to(flights) %>% select(-carrier) %>% glimpse()
#> dm of 5 tables: `airlines`, `airports`, `flights`, `planes`, `weather`
#> 
#> Table: `flights`
#> 3 outgoing foreign key(s):
#>   `origin` -> `airports$faa` no_action
#>   `tailnum` -> `planes$tailnum` no_action
#>   (`origin`, `time_hour`) -> (`weather$origin`, `weather$time_hour`) no_action
#> Rows: 1,761
#> Columns: 18
#> $ year           <int> 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2…
#> $ month          <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1…
#> $ day            <int> 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,…
#> $ dep_time       <int> 3, 16, 450, 520, 530, 531, 535, 546, 549, 550, 553, 553…
#> $ sched_dep_time <int> 2359, 2359, 500, 525, 530, 540, 540, 600, 600, 600, 600…
#> $ dep_delay      <dbl> 4, 17, -10, -5, 0, -9, -5, -14, -11, -10, -7, -7, -7, -…
#> $ arr_time       <int> 426, 447, 634, 813, 824, 832, 1015, 645, 652, 649, 711,…
#> $ sched_arr_time <int> 437, 444, 648, 820, 829, 850, 1017, 709, 724, 703, 715,…
#> $ arr_delay      <dbl> -11, 3, -14, -7, -5, -18, -2, -24, -32, -14, -4, -33, -…
#> $ flight         <int> 727, 739, 1117, 1018, 404, 1141, 725, 380, 6055, 2114, …
#> $ tailnum        <chr> "N571JB", "N564JB", "N171US", "N35204", "N815UA", "N5EA…
#> $ origin         <chr> "JFK", "JFK", "EWR", "EWR", "LGA", "JFK", "JFK", "EWR",…
#> $ dest           <chr> "BQN", "PSE", "CLT", "IAH", "IAH", "MIA", "BQN", "BOS",…
#> $ air_time       <dbl> 183, 191, 78, 215, 210, 149, 191, 39, 48, 36, 51, 201, …
#> $ distance       <dbl> 1576, 1617, 529, 1400, 1416, 1089, 1576, 200, 229, 184,…
#> $ hour           <dbl> 23, 23, 5, 5, 5, 5, 5, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6,…
#> $ minute         <dbl> 59, 59, 0, 25, 30, 40, 40, 0, 0, 0, 0, 0, 0, 0, 0, 5, 0…
#> $ time_hour      <dttm> 2013-01-10 23:00:00, 2013-01-10 23:00:00, 2013-01-10 0…

dm_nycflights13() %>% dm_zoom_to(flights) %>% select(-origin) %>% glimpse()
#> dm of 5 tables: `airlines`, `airports`, `flights`, `planes`, `weather`
#> 
#> Table: `flights`
#> 2 outgoing foreign key(s):
#>   `carrier` -> `airlines$carrier` no_action
#>   `tailnum` -> `planes$tailnum` no_action
#> Rows: 1,761
#> Columns: 18
#> $ year           <int> 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2…
#> $ month          <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1…
#> $ day            <int> 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,…
#> $ dep_time       <int> 3, 16, 450, 520, 530, 531, 535, 546, 549, 550, 553, 553…
#> $ sched_dep_time <int> 2359, 2359, 500, 525, 530, 540, 540, 600, 600, 600, 600…
#> $ dep_delay      <dbl> 4, 17, -10, -5, 0, -9, -5, -14, -11, -10, -7, -7, -7, -…
#> $ arr_time       <int> 426, 447, 634, 813, 824, 832, 1015, 645, 652, 649, 711,…
#> $ sched_arr_time <int> 437, 444, 648, 820, 829, 850, 1017, 709, 724, 703, 715,…
#> $ arr_delay      <dbl> -11, 3, -14, -7, -5, -18, -2, -24, -32, -14, -4, -33, -…
#> $ carrier        <chr> "B6", "B6", "US", "UA", "UA", "AA", "B6", "B6", "EV", "…
#> $ flight         <int> 727, 739, 1117, 1018, 404, 1141, 725, 380, 6055, 2114, …
#> $ tailnum        <chr> "N571JB", "N564JB", "N171US", "N35204", "N815UA", "N5EA…
#> $ dest           <chr> "BQN", "PSE", "CLT", "IAH", "IAH", "MIA", "BQN", "BOS",…
#> $ air_time       <dbl> 183, 191, 78, 215, 210, 149, 191, 39, 48, 36, 51, 201, …
#> $ distance       <dbl> 1576, 1617, 529, 1400, 1416, 1089, 1576, 200, 229, 184,…
#> $ hour           <dbl> 23, 23, 5, 5, 5, 5, 5, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6,…
#> $ minute         <dbl> 59, 59, 0, 25, 30, 40, 40, 0, 0, 0, 0, 0, 0, 0, 0, 5, 0…
#> $ time_hour      <dttm> 2013-01-10 23:00:00, 2013-01-10 23:00:00, 2013-01-10 0…
```

<sup>Created on 2022-06-29 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

TSchiefer replied:
Looks great to me now, thank you!

IndrajeetPatil commented:
Btw, the current output looks quite busy.

Should we introduce a line break between FK/PK information and the table glimpse output?

<img width="1200" alt="Screenshot 2022-06-29 at 09 28 34" src="https://user-images.githubusercontent.com/11330453/176378012-c9abd64f-12eb-4ae0-8d19-ed06453e7ea1.png">

TSchiefer commented:
That is a good point. 
On the other hand, that way all the info for each individual table is contained in one "paragraph". No strong opinion here though.

IndrajeetPatil commented:
Okay, I will post an example output.

Also, when the table is zoomed, shouldn't the output say 

> Zoomed table: `flights`

 instead of 

> Table: `flights`

TSchiefer commented:
Great, thanks.

And absolutely, it should be `Zoomed table: ` then, thanks.

TSchiefer commented:
Thanks, Indra, and sorry for the delayed response.
Short answer: I like it for `zoomed_dm` but not so much for `dm`. For the latter I think the eye is a bit misguided by the newlines, meaning it's a bit harder to tell with a quick glance where one table ends and the other begins.
In case of the method for `dm` we could introduce another separator though, something like `---` (could even be extended to `width`) or something else.

IndrajeetPatil commented:
How about we use a ruler (of green color?) instead? 
We use it everywhere else, so it will be good to be stylistically consistent.

# `dm`

``` r
library(dm)

dm_nycflights13() %>% glimpse()
#> dm of 5 tables: `airlines`, `airports`, `flights`, `planes`, `weather`
#> 
#> ────────────────────────────────────────────────────────────────────────────────
#>
#> Table: `airlines`
#> Primary key: `carrier`
#> 
#> Rows: 15
#> Columns: 2
#> $ carrier <chr> "9E", "AA", "AS", "B6", "DL", "EV", "F9", "FL", "HA", "MQ", "U…
#> $ name    <chr> "Endeavor Air Inc.", "American Airlines Inc.", "Alaska Airline…
#> 
#> ────────────────────────────────────────────────────────────────────────────────
#> 
#> Table: `airports`
#> Primary key: `faa`
#> 
#> Rows: 86
#> Columns: 8
#> $ faa   <chr> "ALB", "ATL", "AUS", "BDL", "BHM", "BNA", "BOS", "BTV", "BUF", "…
#> $ name  <chr> "Albany Intl", "Hartsfield Jackson Atlanta Intl", "Austin Bergst…
#> $ lat   <dbl> 42.74827, 33.63672, 30.19453, 41.93889, 33.56294, 36.12447, 42.3…
#> $ lon   <dbl> -73.80169, -84.42807, -97.66989, -72.68322, -86.75355, -86.67819…
#> $ alt   <dbl> 285, 1026, 542, 173, 644, 599, 19, 335, 724, 778, 146, 236, 1228…
#> $ tz    <dbl> -5, -5, -6, -5, -6, -6, -5, -5, -5, -8, -5, -5, -5, -5, -5, -5, …
#> $ dst   <chr> "A", "A", "A", "A", "A", "A", "A", "A", "A", "A", "A", "A", "A",…
#> $ tzone <chr> "America/New_York", "America/New_York", "America/Chicago", "Amer…
#> 
#> ────────────────────────────────────────────────────────────────────────────────
#> 
#> Table: `flights`
#> 4 outgoing foreign key(s):
#>   `carrier` -> `airlines$carrier` no_action
#>   `origin` -> `airports$faa` no_action
#>   `tailnum` -> `planes$tailnum` no_action
#>   (`origin`, `time_hour`) -> (`weather$origin`, `weather$time_hour`) no_action
#> 
#> Rows: 1,761
#> Columns: 19
#> $ year           <int> 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2…
#> $ month          <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1…
#> $ day            <int> 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,…
#> $ dep_time       <int> 3, 16, 450, 520, 530, 531, 535, 546, 549, 550, 553, 553…
#> $ sched_dep_time <int> 2359, 2359, 500, 525, 530, 540, 540, 600, 600, 600, 600…
#> $ dep_delay      <dbl> 4, 17, -10, -5, 0, -9, -5, -14, -11, -10, -7, -7, -7, -…
#> $ arr_time       <int> 426, 447, 634, 813, 824, 832, 1015, 645, 652, 649, 711,…
#> $ sched_arr_time <int> 437, 444, 648, 820, 829, 850, 1017, 709, 724, 703, 715,…
#> $ arr_delay      <dbl> -11, 3, -14, -7, -5, -18, -2, -24, -32, -14, -4, -33, -…
#> $ carrier        <chr> "B6", "B6", "US", "UA", "UA", "AA", "B6", "B6", "EV", "…
#> $ flight         <int> 727, 739, 1117, 1018, 404, 1141, 725, 380, 6055, 2114, …
#> $ tailnum        <chr> "N571JB", "N564JB", "N171US", "N35204", "N815UA", "N5EA…
#> $ origin         <chr> "JFK", "JFK", "EWR", "EWR", "LGA", "JFK", "JFK", "EWR",…
#> $ dest           <chr> "BQN", "PSE", "CLT", "IAH", "IAH", "MIA", "BQN", "BOS",…
#> $ air_time       <dbl> 183, 191, 78, 215, 210, 149, 191, 39, 48, 36, 51, 201, …
#> $ distance       <dbl> 1576, 1617, 529, 1400, 1416, 1089, 1576, 200, 229, 184,…
#> $ hour           <dbl> 23, 23, 5, 5, 5, 5, 5, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6,…
#> $ minute         <dbl> 59, 59, 0, 25, 30, 40, 40, 0, 0, 0, 0, 0, 0, 0, 0, 5, 0…
#> $ time_hour      <dttm> 2013-01-10 23:00:00, 2013-01-10 23:00:00, 2013-01-10 0…
#> 
#> ────────────────────────────────────────────────────────────────────────────────
#> 
#> Table: `planes`
#> Primary key: `tailnum`
#> 
#> Rows: 945
#> Columns: 9
#> $ tailnum      <chr> "N10156", "N104UW", "N10575", "N105UW", "N110UW", "N11106…
#> $ year         <int> 2004, 1999, 2002, 1999, 1999, 2002, 2002, 2002, 2003, 200…
#> $ type         <chr> "Fixed wing multi engine", "Fixed wing multi engine", "Fi…
#> $ manufacturer <chr> "EMBRAER", "AIRBUS INDUSTRIE", "EMBRAER", "AIRBUS INDUSTR…
#> $ model        <chr> "EMB-145XR", "A320-214", "EMB-145LR", "A320-214", "A320-2…
#> $ engines      <int> 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, …
#> $ seats        <int> 55, 182, 55, 182, 182, 55, 55, 55, 55, 55, 55, 55, 55, 55…
#> $ speed        <int> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N…
#> $ engine       <chr> "Turbo-fan", "Turbo-fan", "Turbo-fan", "Turbo-fan", "Turb…
#> 
#> ────────────────────────────────────────────────────────────────────────────────
#> 
#> Table: `weather`
#> Primary key: (`origin`, `time_hour`)
#> 
#> Rows: 144
#> Columns: 15
#> $ origin     <chr> "EWR", "EWR", "EWR", "EWR", "EWR", "EWR", "EWR", "EWR", "EW…
#> $ year       <int> 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013,…
#> $ month      <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,…
#> $ day        <int> 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,…
#> $ hour       <int> 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 1…
#> $ temp       <dbl> 41.00, 39.02, 39.02, 39.92, 41.00, 41.00, 39.92, 41.00, 42.…
#> $ dewp       <dbl> 32.00, 30.02, 28.94, 26.96, 26.06, 26.06, 24.98, 24.98, 24.…
#> $ humid      <dbl> 70.08, 69.86, 66.85, 59.50, 54.97, 54.97, 54.81, 52.56, 48.…
#> $ wind_dir   <dbl> 230, 210, 230, 270, 320, 300, 280, 330, 330, 320, 320, 330,…
#> $ wind_speed <dbl> 8.05546, 9.20624, 6.90468, 5.75390, 6.90468, 12.65858, 6.90…
#> $ wind_gust  <dbl> NA, NA, NA, NA, NA, 20.71404, 17.26170, NA, NA, 26.46794, N…
#> $ precip     <dbl> 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,…
#> $ pressure   <dbl> 1024.6, 1025.9, 1026.9, 1027.5, 1028.2, 1029.0, 1030.0, 103…
#> $ visib      <dbl> 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,…
#> $ time_hour  <dttm> 2013-01-10 00:00:00, 2013-01-10 01:00:00, 2013-01-10 02:00…
#> 
#> ────────────────────────────────────────────────────────────────────────────────
```

# `zoomed_dm`

```r
dm_nycflights13() %>% dm_zoom_to(weather) %>% glimpse()
#> dm of 5 tables: `airlines`, `airports`, `flights`, `planes`, `weather`
#> 
#> Zoomed table: `weather`
#> Primary key: (`origin`, `time_hour`)
#> 
#> Rows: 144
#> Columns: 15
#> $ origin     <chr> "EWR", "EWR", "EWR", "EWR", "EWR", "EWR", "EWR", "EWR", "EW…
#> $ year       <int> 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013,…
#> $ month      <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,…
#> $ day        <int> 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,…
#> $ hour       <int> 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 1…
#> $ temp       <dbl> 41.00, 39.02, 39.02, 39.92, 41.00, 41.00, 39.92, 41.00, 42.…
#> $ dewp       <dbl> 32.00, 30.02, 28.94, 26.96, 26.06, 26.06, 24.98, 24.98, 24.…
#> $ humid      <dbl> 70.08, 69.86, 66.85, 59.50, 54.97, 54.97, 54.81, 52.56, 48.…
#> $ wind_dir   <dbl> 230, 210, 230, 270, 320, 300, 280, 330, 330, 320, 320, 330,…
#> $ wind_speed <dbl> 8.05546, 9.20624, 6.90468, 5.75390, 6.90468, 12.65858, 6.90…
#> $ wind_gust  <dbl> NA, NA, NA, NA, NA, 20.71404, 17.26170, NA, NA, 26.46794, N…
#> $ precip     <dbl> 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,…
#> $ pressure   <dbl> 1024.6, 1025.9, 1026.9, 1027.5, 1028.2, 1029.0, 1030.0, 103…
#> $ visib      <dbl> 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,…
#> $ time_hour  <dttm> 2013-01-10 00:00:00, 2013-01-10 01:00:00, 2013-01-10 02:00…
```

<sup>Created on 2022-07-01 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

TSchiefer commented:
I feel really picky now, but could we just use the default color here? Like it quite a bit apart from that.

IndrajeetPatil commented:
> could we just use the default color here

Default color for `cli::cat_rule()`, right? Done.

TSchiefer commented:
LGTM, thank you!

krlmlr commented:
Thanks!
